### DISCUSSION TITLE
Airflow 2.7 to 2.8 upgrade failure

### DISCUSSION THREAD

**Discussion by AlekseiSaff** [apache/airflow#37027] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.7.3

### What happened?

after succsessfull pip upgrade to 2.8 or 2.8.1 we are running airflow db upgrade and receiving

`[2024-01-26T10:19:46.257+0800] {cli_parser.py:78} WARNING - Authentication manager is not configured and webserver will not be able to start.
/usr/local/lib/python3.9/site-packages/airflow/cli/commands/db_command.py:68 DeprecationWarning: `db upgrade` is deprecated. Use `db migrate` instead.
DB: mysql://airflow:***@localhost/airflow_db
Performing upgrade to the metadata database mysql://airflow:***@localhost/airflow_db
Traceback (most recent call last):
  File "/usr/local/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/airflow/__main__.py", line 57, in main
    args.func(args)
  File "/usr/local/lib/python3.9/site-packages/airflow/cli/cli_config.py", line 49, in command
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/cli/commands/db_command.py", line 69, in upgradedb
    migratedb(args)
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/cli.py", line 114, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/providers_configuration_loader.py", line 55, in wrapped_function
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/cli/commands/db_command.py", line 129, in migratedb
    db.upgradedb(
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/session.py", line 79, in wrapper
    return func(*args, session=session, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/db.py", line 1572, in upgradedb
    import_all_models()
  File "/usr/local/lib/python3.9/site-packages/airflow/models/__init__.py", line 60, in import_all_models
    __getattr__(name)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/__init__.py", line 76, in __getattr__
    val = import_string(f"{path}.{name}")
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/module_loading.py", line 39, in import_string
    module = import_module(module_path)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/usr/local/lib/python3.9/site-packages/airflow/models/dag.py", line 94, in <module>
    from airflow.jobs.job import run_job
  File "/usr/local/lib/python3.9/site-packages/airflow/jobs/job.py", line 35, in <module>
    from airflow.serialization.pydantic.job import JobPydantic
  File "/usr/local/lib/python3.9/site-packages/airflow/serialization/pydantic/job.py", line 33, in <module>
    class JobPydantic(BaseModelPydantic):
  File "pydantic/main.py", line 222, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 506, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 436, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 546, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 570, in pydantic.fields.ModelField._set_default_and_type
  File "pydantic/fields.py", line 439, in pydantic.fields.ModelField.get_default
  File "pydantic/utils.py", line 695, in pydantic.utils.smart_deepcopy
  File "/usr/local/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/local/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/local/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/local/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/local/lib/python3.9/copy.py", line 161, in deepcopy
    rv = reductor(4)
TypeError: cannot pickle '_thread.RLock' object
[lt@lt-hkg1-avm04 experiments]$ 

`

### What you think should happen instead?

DB should be updated to 2.8.1

### How to reproduce

upgrade 2.7 to 2.8 via PIP and run airflow db upgrade

### Operating System

CentOS 7.9

### Deployment

Other

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Please use constraints when you upgrade https://airflow.apache.org/docs/apache-airflow/stable/installation/index.html#using-pypi

**Comment by AlekseiSaff** [comment]
Of course, I did. I'm maintaining that airflow instance from 1.x to 2.x and updating to all minor versions
`pip3.9 install apache-airflow[celery,mysql,slack]==2.8.1 --upgrade --constraint https://raw.githubusercontent.com/apache/airflow/constraints-2.8.1/constraints-3.9.txt
`

**Reply by potiuk** [reply]
What's your pydantic version ?

**Comment by AlekseiSaff** [comment]
Airflow 2.7.3
>>> print (pydantic.__version__)
2.4.2


pip3.9 install apache-airflow[celery,mysql,slack]==2.8.1 --upgrade --constraint https://raw.githubusercontent.com/apache/airflow/constraints-2.8.1/constraints-3.9.txt 



Successfully installed Babel-2.14.0 Flask-JWT-Extended-4.6.0 Mako-1.3.0 WTForms-3.1.2 aiohttp-3.9.1 alembic-1.13.1 amqp-5.2.0 anyio-4.2.0 apache-airflow-2.8.1 apache-airflow-providers-celery-3.5.1 apache-airflow-providers-common-io-1.2.0 apache-airflow-providers-common-sql-1.10.0 apache-airflow-providers-ftp-3.7.0 apache-airflow-providers-http-4.8.0 apache-airflow-providers-imap-3.5.0 apache-airflow-providers-mysql-5.5.1 apache-airflow-providers-slack-8.5.1 apache-airflow-providers-sqlite-3.7.0 apispec-6.4.0 argcomplete-3.2.1 attrs-23.2.0 backoff-2.2.1 billiard-4.2.0 blinker-1.7.0 celery-5.3.6 certifi-2023.11.17 configupdater-3.2 cryptography-41.0.7 exceptiongroup-1.2.0 flask-appbuilder-4.3.10 frozenlist-1.4.1 fsspec-2023.12.2 googleapis-common-protos-1.62.0 greenlet-3.0.3 grpcio-1.60.0 humanize-4.9.0 idna-3.6 importlib-metadata-6.11.0 importlib-resources-6.1.1 jinja2-3.1.3 jsonschema-4.20.0 jsonschema-specifications-2023.12.1 kombu-5.3.5 lazy-object-proxy-1.10.0 limits-3.7.0 markdown-3.5.2 marshmallow-3.20.2 mysqlclient-2.2.1 opentelemetry-api-1.22.0 opentelemetry-exporter-otlp-1.22.0 opentelemetry-exporter-otlp-proto-common-1.22.0 opentelemetry-exporter-otlp-proto-grpc-1.22.0 opentelemetry-exporter-otlp-proto-http-1.22.0 opentelemetry-proto-1.22.0 opentelemetry-sdk-1.22.0 opentelemetry-semantic-conventions-0.43b0 pathspec-0.12.1 pendulum-3.0.0 prometheus-client-0.19.0 prompt-toolkit-3.0.43 protobuf-4.25.2 psutil-5.9.7 pydantic-2.5.3 pydantic-core-2.14.6 pygments-2.17.2 referencing-0.32.1 rich-13.7.0 rpds-py-0.17.1 slack_sdk-3.26.2 sqlalchemy-1.4.51 sqlalchemy-jsonfield-1.0.2 termcolor-2.4.0 time-machine-2.13.0 tornado-6.4 typing-extensions-4.9.0 tzdata-2023.4 universal-pathlib-0.1.4 vine-5.1.0 wcwidth-0.2.13 wrapt-1.16.0 yarl-1.9.4


>>> import pydantic
>>> print (pydantic.__version__)
2.5.3

**Reply by potiuk** [reply]
Is this issue persistent now? Do you have some special configuration changed comparing to plain Airflow ? Do you use pickling ? What happens when you try it on a clean airflow installation?

The reason is - we cannot reproduce it, you have to try to attempt to see what triggers the error for you, because without reprodcuing it, we can't fix it.

**Reply by potiuk** [reply]
Anything, you deviated from standard airflow? Any modification in any settings configuration, any special customizations?

**Reply by potiuk** [reply]
Also `db upgrade` is (as you see) deprecated. What happens when you use `db migrate` as suggested? Do you have any other deprecation warning that you have not complied with ?

**Reply by potiuk** [reply]
For some reason Auth Manager also seems to be not configured, so that suggests, that your environment is somewhat broken at least with the configuration. This would only happen if you use latest main and have no `fab` provider installed  (which is coming in Airflow 2.9),  or if you deliberately changed your configuration to point to non-existing or missing Authentication Manager, so  this suggest you have a deeper problem with your environment.

**Reply by potiuk** [reply]
I suggest to install Airflow (2.8.0 or 2.8.1)  in a separate, new environment  and try again. There is generally no particular reason to upgrade airflow in-place. If you are using virtualenv, it's also perfect fine to have several venvs separate from each other and activate and use the one you need to use at a given moment.

**Reply by potiuk** [reply]
Let me know results of your experiment.

**Reply by AlekseiSaff** [reply]
1. that is the virtual machine, so I can do install revert back to previous version. So now I again reverted back to 2.7.3. 
2. Nothing special, that is a long-term airflow (more then 4 years, which was always following new versions update)
3, db migrate provides the same error

as for providers, that's what I see in WEB GUI list

<h2 style="box-sizing: border-box; font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-weight: 500; line-height: 1.1; color: rgb(81, 80, 79); margin-top: 20px; margin-bottom: 10px; font-size: 30px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Providers</h2>
Package Name | Version | Description
-- | -- | --
apache-airflow-providers-amazon | 6.0.0 | Amazon integration (including Amazon Web Services (AWS)).
apache-airflow-providers-celery | 3.4.1 | Celery
apache-airflow-providers-common-sql | 1.8.0 | Common SQL Provider
apache-airflow-providers-ftp | 3.6.0 | File Transfer Protocol (FTP)
apache-airflow-providers-http | 4.6.0 | Hypertext Transfer Protocol (HTTP)
apache-airflow-providers-imap | 3.4.0 | Internet Message Access Protocol (IMAP)
apache-airflow-providers-mysql | 5.4.0 | MySQL
apache-airflow-providers-slack | 8.3.0 | Slack
apache-airflow-providers-sqlite | 3.5.0 | SQLite

Providers
Package Name	Version	Description
[apache-airflow-providers-amazon](https://airflow.apache.org/docs/apache-airflow-providers-amazon/6.0.0/)	6.0.0	Amazon integration (including [Amazon Web Services (AWS)](https://aws.amazon.com/)).
[apache-airflow-providers-celery](https://airflow.apache.org/docs/apache-airflow-providers-celery/3.4.1/)	3.4.1	[Celery](http://www.celeryproject.org/)
[apache-airflow-providers-common-sql](https://airflow.apache.org/docs/apache-airflow-providers-common-sql/1.8.0/)	1.8.0	[Common SQL Provider](https://en.wikipedia.org/wiki/SQL)
[apache-airflow-providers-ftp](https://airflow.apache.org/docs/apache-airflow-providers-ftp/3.6.0/)	3.6.0	[File Transfer Protocol (FTP)](https://tools.ietf.org/html/rfc114)
[apache-airflow-providers-http](https://airflow.apache.org/docs/apache-airflow-providers-http/4.6.0/)	4.6.0	[Hypertext Transfer Protocol (HTTP)](https://www.w3.org/Protocols/)
[apache-airflow-providers-imap](https://airflow.apache.org/docs/apache-airflow-providers-imap/3.4.0/)	3.4.0	[Internet Message Access Protocol (IMAP)](https://tools.ietf.org/html/rfc3501)
[apache-airflow-providers-mysql](https://airflow.apache.org/docs/apache-airflow-providers-mysql/5.4.0/)	5.4.0	[MySQL](https://www.mysql.com/products/)
[apache-airflow-providers-slack](https://airflow.apache.org/docs/apache-airflow-providers-slack/8.3.0/)	8.3.0	[Slack](https://slack.com/)
[apache-airflow-providers-sqlite](https://airflow.apache.org/docs/apache-airflow-providers-sqlite/3.5.0/)	3.5.0	[SQLite](https://www.sqlite.org/)

**Comment by AlekseiSaff** [comment]
I've found couple of similar issues
https://stackoverflow.com/questions/48933382/airflow-1-9-0-externaltasksensor-retry-delay-30-yields-typeerror-cant-pickle

it seems like auth_manager is only WARNING signal

**Reply by potiuk** [reply]
`asked 5 years ago` Airflow 1.9.0. The message is a general message that signals that you try to serialize something that is not serializable - for example badly written logger. There is almost no code in Airflow 1.9 vs now. It has nothing to do with your message  (unless you have some unknown and bad configuration from Airflow 1.9 that has been caried over for many years - where you (possibly) have not solved similar deprecation warnings.

Again:

* start from scratch (including generating configuration from scratch)
* make sure that you review your past configuration and apply the configuration you have one-by-one deliberately, knowing what you are doing (not just by copying airflow.cfg/env variables)
* make sure your all deprecation warning

If you see it happening again, copy the CHANGES in the configuration that are triggering it so that we are able to reproduce it

This is the only way I can advice you to proceed.

**Reply by AlekseiSaff** [reply]
I tried to play with many parameters and import 
auth_manager = airflow.auth.managers.fab.fab_auth_manager.FabAuthManager

which seems is not necessary as this is defauly value. So basically, two WARNING/ERROR lines

Jan 28 18:33:34 lt-hkg1-avm04 airflow: [#033[34m2024-01-28T18:33:34.246+0800#033[0m] {#033[34mcli_parser.py:#033[0m77} WARNING#033[0m - cannot load CLI commands from auth manager: can
not pickle '_thread.RLock' object#033[0m
Jan 28 18:33:34 lt-hkg1-avm04 airflow: [#033[34m2024-01-28T18:33:34.248+0800#033[0m] {#033[34mcli_parser.py:#033[0m78} WARNING#033[0m - Authentication manager is not configured and we
bserver will not be able to start.#033[0m
Jan 28 18:33:34 lt-hkg1-avm04 airflow: [#033[34m2024-01-28T18:33:34.248+0800#033[0m] {#033[34mcli_parser.py:#033[0m81} ERROR#033[0m - cannot pickle '_thread.RLock' object#033[0m


So I'm not sure what I need to configure as we have in webserver_config.py
AUTH_TYPE = AUTH_LDAP

**Reply by potiuk** [reply]
> {https://github.com/apache/airflow/pull/33[34mcli_parser.py:https://github.com/apache/airflow/pull/33[0m77} WARNING#033[0m - cannot load CLI commands from auth manager: can
not pickle '_thread.RLock' object#033[0m

You seem to have some generic problem where pickling is happening even when you are retrieving the auth manager CLI commands ad get exactly the same pickle command. There is something broken in your environment. Did you install it completely from scratch in a separate virtualenv ? If so, how ?

**Reply by potiuk** [reply]
I think also what COULD have happened is that you could have your venv generated in different environment and copied to your environment from elsewhere. For example if you generate your environment in your host, and then you try to mount it to your docker container.

Are you doing something like that and forgot to mention this ?

**Reply by AlekseiSaff** [reply]
sorry, I'm not using any docker/venv, etc. That is just regular host with CentOS 7.9

**Reply by potiuk** [reply]
So my guess is that you are using broken Python that is coming with that Centos. Please install Python from regular Python sources. There are known issues where Python delivered by the distro is broken because distro developers modified it. That would be my guess. Or even you attempt to install Airflow directly using Python that comes with the browser without creating venv, which is even worse and strongly discouraged by everyone (including warnings printed by `pip` when you attempt doing it. ALWAYS USE VENV (unless you are installing airflow in container for in-container usage).

Install Python from here https://www.python.org/downloads/ and recreate the venv. BTW. you have not responed if you recreated the venv already?  How did you do it ?

Did I miss that answer?

**Reply by AlekseiSaff** [reply]
we are using python 3.9 from regular sources, as the default in centos is 3.6
I mentioned previously, that we are not using venv/docker

**Reply by potiuk** [reply]
Well then use venv and create it from the scratch.

**Reply by potiuk** [reply]
And just to add more context - It's a very bad practice to install any Python application directly without venv (unless in container) because you can't do what I told you - se use venvs, to switch between different versions and because system level packages are interfering with the ones you install in the application 

There were long dicsussions about it in Python Packaging team resulting in https://peps.python.org/pep-0668/ and finally into this docuement where they explain the problems when you do it and stronly recommend both users to use venv and distro maintainers to guide users to use venvs to install their application  https://packaging.python.org/en/latest/specifications/externally-managed-environments/#externally-managed-environments

You might want to take a look at those docs.

**Reply by potiuk** [reply]
And let me know if the problem re-occurs when you use venv. Would be curious to see it.

**Reply by AlekseiSaff** [reply]
am I right in saying:
 - I need to install 2.8.1 in venv
 - I need to use the existing DB and point new 2.8.1 to that DB/rabbitMQ
 - I need to use the exisitng conf file for webserver/scheduler and point new 2.8.1 to that config
 
 ?

**Reply by AlekseiSaff** [reply]
The whole current machine is dedicated to airflow, and as I mentioned, I never ever had problems since 1.x till current 2.7.3 (doing update all version 1 > 2.0 > 2.1> 2.2>....2.6->2.7

**Reply by potiuk** [reply]
I understand, still installing airflow in a separate venv is  how it **should** be done. 

> I need to install 2.8.1 in venv

Yes.

> I need to use the existing DB and point new 2.8.1 to that DB/rabbitMQ

Yes

>  need to use the exisitng conf file for webserver/scheduler and point new 2.8.1 to that config

Ideally - as explained before - start with the default configuration of Airflow and deliberately set configuration parameters that you want to change one-by-one comparing to the defaults. This will help you to get rid of the deprecation warnings as well, and you will have a chance to review if you do mWe recommend a relatively new way of doing it where you can produce a default config by running `airflow config list --defaults` which will generate config file with all values explained and commented out and you will be able to set the configuration in this file by looking at your old configuration and deliberately deciding if you want to do anything about it. 

BTW. If you find - during that review - that at any point you used `pickling` in some of the parameters - do not enable that. This has been default in 1.10 in some configuration and since you seem to go up from 1.10 you probably accumulated there some changes that are not defaults and that might induce that behaviour. That might also mean that your plckled data in the database needs to be cleaned - but this is a manual process, so I hope it's not the case.

**Reply by potiuk** [reply]
See https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html#setting-configuration-options

**Reply by AlekseiSaff** [reply]
since it's a new week and airflow is under heavy load, I'll try to do more experiments next weekend.
so far: nothing about pickling in the config

and big news for me is that we need to rewrite the config for every upgrade. so far not a big issue for me as we can stay on 2.7.3, just don't want to be outdated at one moment in time

**Reply by potiuk** [reply]
no - you do not need to rewrite for every update. But it's lkely, if you did not fix deprecation warnings before (and we know you did not) - and you brought your configuration from 1.10, it might have problems. If you follow the approach I wrote - where you start from commented out defaults and only enable /modify the configuration you need, you will be good in keeping the configuraiton updated (as lnog as you also fix the deprecation warnings(

We are trying to help you to solve YOUR problem in the way that will get YOUR installation back to the shape and - responding to your requests, we are trying to give you hints and ideas how you can - in the fastest way bring it to the state that will be good for it. You can follow it, or contest it's up to you. We are doing it in our free time and try to help you to configure it without losing to much ours and your time.

I suggest you simply take advantage of it and simply follow good practices : creating venv, recrating configuration having separate installation for different versions. Those ar things that are known to be good practices and work. We have no crystal ball to know what you have done in your environment in the past that might have caused it and no time of everyone to try to reverse engineer it, so we are simply suggesting what is the fastest and most robust way for you to make it works.

It's an advice, you can do with what you want - follow it or not, follow best practices or not, if you follow them, it narrows down the problems and helps us to help you ,ore effectively, if you don't then at some point in time you are left on your own with solving your own problems.

**Reply by AlekseiSaff** [reply]
and we know you did not - that is not true. every time I remove / update configs, no deprecation warnings are published, I can show you log os 2.7.3 startup and you will see there is NO deprecation warnings.

I will try, but as I said, airflow is busy right now, so I can't do it during production hours. 

and anyway always thanks for your time / help / suggestions!

**Comment by bolkedebruin** [comment]
I have only seen `TypeError: cannot pickle '_thread.RLock' object` with pydantic needing an update/change. Support for `cached_property` (computed_fields) has been flaky and the update to support it is quite recent. I suggest:

* see if an update of pydantic-core helps
* upgrade to a newer version of python (3.10+)

**Reply by AlekseiSaff** [reply]
pydantic and pydantic-core are part of the constraints.
python 3.9 is still under support

**Reply by bolkedebruin** [reply]
I know they are, but you need to change something. This might be a case that these issues were not found during testing and hence not fixed by the constraints. You can always downgrade again if it doesn't change anything. 

Let us know when you upgraded / tried a change as we cannot reproduce it on our side thus not debug it.

**Comment by AlekseiSaff** [comment]
@potiuk just to let you know that the same setup update to 2.8.2 is successful, when 2.8.0 and 2.8.1 both broke everything with an error I posted above

**Reply by potiuk** [reply]
Aha. OK. Then the problem was connected with Pydantic probably. In Airlfow 2.8.2 we made it an optional dependency and that probably helpes. Probably you had some Pydantic-related dependencies that triggered it.

**Reply by AlekseiSaff** [reply]
do you think it's related to my DAGs?

**Reply by potiuk** [reply]
Could be. Not sure.
